thread_primitives.h

```diff
@@ -28,65 +28,119 @@
-    Mutex* mutex_;
+public:
+  ScopedLock(Mutex *mutex) : mutex_(mutex) { mutex->lock(); }
+  ~ScopedLock() { mutex_->unlock(); }
+
+private:
+  ScopedLock(const ScopedLock &) {}
+  ScopedLock &operator=(const ScopedLock &) { return *this; }
+  Mutex *mutex_;
 };
 
 class ConditionVariable {
-  public:
-    ConditionVariable();
-    ~ConditionVariable();
-    
-    void signal();
-    void wakeAll();
-    void wait(Mutex* mutex);
-
-  private:
-    ConditionVariablePrivate* impl_;
+public:
+  ConditionVariable() = default;
+
+  void signal() { _condvar.notify_one(); }
+  void wakeAll() { _condvar.notify_all(); }
+  void wait(Mutex *mutex) {
+    std::unique_lock<std::mutex> ulock(mutex->mutex_, std::defer_lock);
+    _condvar.wait(ulock);
+  }
+
+private:
+  std::condition_variable _condvar;
 };
 
 class Thread {
-  public:
-    Thread();
-    ~Thread();
+public:
+  Thread() = default;
+  ~Thread() {
+    if (thread_.joinable()) {
+      thread_.detach();
+    }
+  }
 
-    bool start(int (*func)(void *), void *ptr);
+  bool start(int (*func)(void *), void *ptr) {
 
-    bool isRunning() const;
+    if (isRunning()) {
+      return false;
+    }
 
-    // Returns the thread exit value, as returned by 'func' above.
-    // returns -1 if the thread has never been started.
-    int waitForTermination();
+    // we may need to wait for the thread to finish, as it may be still
+    // terminating (flag running_ can be set before thread ends)
+    if (thread_.joinable())
+      thread_.join();
 
-    static void setCurrentName(const char *name);
+    // Wrap the function in a lambda to signal when the thread has finished
+    // processing
+    std::packaged_task<int(void *, std::weak_ptr<bool>)> task(
+        [func](void *p, std::weak_ptr<bool> runningFlag) {
+          const int retVal = func(p);
+          if (auto r = runningFlag.lock()) {
+            *r = false;
+          }
 
-  private:
-    void* thread_handle;
-};
+          return retVal;
+        });
+    thread_ret_future_ = task.get_future();
+
+    *running_ = true;
+    thread_ =
+        std::thread(std::move(task), ptr, std::weak_ptr<bool>(this->running_));
 
+    return true;
+  }
+
+  bool isRunning() const { return *running_; }
+
+  // Returns the thread exit value, as returned by 'func' above.
+  // returns -1 if the thread has never been started.
+  int waitForTermination() {
+    if (thread_.joinable()) {
+      thread_.join();
+    }
+
+    // make sure we call .get() once. If called > 1 time it will throw an
+    // exception
+    return thread_ret_future_.valid() ? thread_ret_future_.get() : -1;
+  }
+
+  static void setCurrentName(const char *name) { (void)name; }
+
+private:
+  std::thread thread_;
+  std::future<int>
+      thread_ret_future_; // holds future return value of thread fn.
+
+  // the thread may outlive this class (see destructor's detach()), so we use a
+  // shared_ptr here
+  std::shared_ptr<bool> running_ = std::make_shared<bool>(false);
```

Comment:
finding out if a thread is running or finished is not easy with std::thread. I have a workaround here that it a bit more complicated than what I'd like. Let me know if you have other suggestions.

Reply:
no need for that. thread_.joinable() will do the job better.